compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -1160,4 +1160,34 @@ object messages {
            |                           ^^^^
            |""".stripMargin
   }
+
+  case class OverloadedOrRecursiveMethodNeedsResultType(tree: Names.TermName)(implicit ctx: Context)
+  extends Message(OverloadedOrRecursiveMethodNeedsResultTypeID) {
+    val kind = "Syntax"
+    val msg = hl"""overloaded or recursive method ${tree} needs result type"""
+    val explanation =
+      hl"""|${tree} is overloaded and at least one definition of it calls another.
+           |You need to specify the calling method's return type.
+         """.stripMargin
```

Comment:
In this case the method *might* be overloaded but it could also be simply recursive. So claiming that it is overloaded is false. We might try to find out whether it is overloaded to improve the error message. Essentially, it's overloaded if its owner is a class and taking the member of the name in the owner gives several alternatives.